src/main/java/com/hivemq/mqtt/handler/publish/PublishSendHandler.java

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.mqtt.handler.publish;
+
+import com.codahale.metrics.Counter;
+import com.hivemq.configuration.service.InternalConfigurations;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+import com.hivemq.metrics.MetricsHolder;
+import com.hivemq.mqtt.message.publish.PublishWithFuture;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * @author Daniel Krüger
+ */
+public class PublishSendHandler extends ChannelInboundHandlerAdapter implements Runnable {
+
+    private final @NotNull LinkedList<PublishWithFuture> messagesToWrite = new LinkedList<>();
+    private final @NotNull Counter channelNotWritable;
+    private @Nullable ChannelHandlerContext ctx;
+    private boolean wasWritable = true;
+
+    public PublishSendHandler(final @NotNull MetricsHolder metricsHolder) {
+        channelNotWritable = metricsHolder.getChannelNotWritableCounter();
+    }
+
+    @Override
+    public void run() {
+        consumeQueue();
+    }
+
+    @Override
+    public void handlerAdded(final @NotNull ChannelHandlerContext ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public void channelWritabilityChanged(final @NotNull ChannelHandlerContext ctx) {
+        final Channel channel = ctx.channel();
+        if (channel.isWritable()) {
+            if (!wasWritable) {
+                wasWritable = true;
+                channelNotWritable.dec();
+            }
+
+            channel.eventLoop().execute(this);
+        }
+        ctx.fireChannelWritabilityChanged();
+    }
+
+    public void sendPublishes(final @NotNull List<PublishWithFuture> publishes) {
+        assert ctx != null : "Context must not be null";
+        ctx.channel().eventLoop().execute(() -> {
+            messagesToWrite.addAll(publishes);
+            consumeQueue();
+        });
+    }
+
+    private void consumeQueue() {
+        assert ctx != null : "Context must not be null";
+        int written = 0;
+        while (!messagesToWrite.isEmpty()) {
+            if (!ctx.channel().isWritable()) {
+                if (wasWritable) {
+                    wasWritable = false;
+                    channelNotWritable.inc();
+                }
+                break;
+            }
+            final PublishWithFuture publishFuture = messagesToWrite.poll();
+            ctx.write(publishFuture).addListener(new PublishWriteFailedListener(publishFuture.getFuture()));
+            written++;
+            if (written >= InternalConfigurations.MAX_PUBLISHES_BEFORE_FLUSH.get()) {
```

Comment:
always reading the atomic integer here, seems wasteful

Reply:
Can `InternalConfigurations` be changed dynamically after startup?
If no, I would just move it into the constructor.

Reply:
in unit tests yet, but not in normal cases i guess. Should do no harm to read it in the constructor